Support GDB with native rust support

TimNN opened the pull request:
This PR aims at making the debuginfo tests pass with the newer GDB versions which have native rust support (RGDB)

To that end debuginfo tests now support three GDB prefixes:
- `gdb` applicable to both GDB varieties
- `gdbg` (**G**eneric) only applicable to the old GDB
- `gdbr` (**R**ust) only applicable to the new RGDB

Whether the GDB has rust support is detected based on it's version: GDB >= 7.11.10 is considered to have rust support.

---

**Test updates**

All tests have been updated to the new GDB version. Note that some tests currently require the GDB trunk<sup>1</sup>.

---

**How to move forward with this PR:**

I propose the following steps for moving forward with this PR:
- [x] Validate the general approach of this PR (the `gdb-`, `gdbg-` and `gdbr-` split)
- [x] Validate the approach taken for updating the debuginfo tests (I've checked this since there's (almost) no `set language c` left, which was my main concern)
- [x] Determine how to distinguish between the new and old GDB (and implement that)
- [ ] Add one or more non-gating build bots with the new GDB (blocked on the previous item, can happen after this PR has been merged)
- [ ] If the new bots pass the tests, gate on them
- [x] \(Maybe) update the remaining tests to run without `set syntax c` (in a separate PR)
- [ ] \(Maybe) add tests specifically for the new GDB (in a separate PR / open an issue about this)

I'm not completely sure about the build bot related steps (cc @alexcrichton), the current approach was suggested to prevent any downtime / broken build time between a new GDB gating builder being added and this PR being merged.

---

**Suboptimal RGDB Output**

I've found several places where the output of RGDB is not ideal. They are tagged with `// FIXME`, here is an overview:
- [x] Trait references are not syntactically correct: `type_names::&Trait2<...>` (**WontFix**: the issue is minor and from @Manishearth below: "to properly address the trait issue we should wait for trait object support")
- [x] Univariant enums are printed as `<error reading variable>` (**Fixed** in GDB trunk<sup>1<sup>)
- [x] Unions are printed as `<error reading variable>` (**Fixed** in GDB trunk<sup>1</sup>)
- [x] "Nil Enums" (`enum Foo {}`) are printed as `<error reading variable>` (**WontFix**: the are not supposed to exist)
- [x] I have found no alternative for `sizeof(var)` in rust mode, so had to resort to `set language c` (**Fixed** in GDB trunk<sup>1</sup>)
- [x] I have found not way of interpreting a value as a specific enum variant, so had to resort to `set language c` (**Fixed** in GDB trunk<sup>1</sup>)
- [x] Before the initial `run` command, gdb does not realise it should be in rust mode (thus, if one want's to print statics before the run one has to explicitly `set language rust`) (maybe this is intended / expected behaviour, if so, someone please tell me ;) (**"Expected"** / current behaviour of GDB: picks up jemalloc, see https://github.com/rust-lang/rust/pull/37410#issuecomment-257068472)

<sup>1</sup>: Or rather in @Manishearth's trunk, waiting to be merged upstream.

---

cc @alexcrichton, @michaelwoerister, #36323

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

TimNN commented:
**How to move forward with this PR:**

I propose the following steps for moving forward with this PR:

* [ ] Validate the general approach of this PR (the `gdb-`, `gdbg-` and `gdbr-` split)
* [ ] Validate the approach taken for updating the debuginfo tests
* [ ] Determine how to distinguish between the new and old GDB (and implement that)
* [ ] Add one or more non-gating build bots with the new GDB (blocked on the previous item, can happen after this PR has been merged)
* [ ] If the new bots pass the tests, gate on them
* [ ] \(Maybe) update the remaining tests to run without `set syntax c` (in a separate PR)
* [ ] \(Maybe) add tests specifically for the new GDB (in a separate PR / open an issue about this)

I'm not completely sure about the build bot related steps (cc @alexcrichton), the current approach was suggested to prevent any downtime / broken build time between a new GDB gating builder being added and this PR being merged.

If anyone wants to discuss this PR more directly, feel free to ping me on irc (TimNN).

michaelwoerister commented:
Oh nice! This has been near the top of my list of things to do for months. I give you more feedback later today.

michaelwoerister commented:
> Validate the general approach of this PR (the gdb-, gdbg- and gdbr- split)

I think we can check this box. This is basically what we planned to do in the tools team and it allows for nice sharing of commands when writing test cases.

michaelwoerister commented:
> Determine how to distinguish between the new and old GDB (and implement that)

There is [some version detection](https://github.com/rust-lang/rust/blob/a6b3b01b5f7f5a9d7d340dacf7dbf72be29e2c07/src/tools/compiletest/src/main.rs#L513) we already do for LLDB via the output of `lldb --version`. What we have learned from this is:
* the output of `--version` can be rather unreliable, because distros and vendors will modify it as they please,
* but that's not much of a problem because there's still a rather limited number of general formats and we can just add logic for those as we come across them.

So I'd say it's definitely possible to go down that road. I'm not sure the feature detection approach is more stable (it depends on how stable the error message is). I'd slightly favor using the same approach we are using for LLDB.

michaelwoerister commented:
Ideally I'd like both versions of GDB to be tested on all gating build bots. We could let `configure` define a `PRE_RUST_GDB` and a `RUST_ENABLED_GDB`, both of which will be used by `compiletest` to run debuginfo tests. `configure` would take explicit arguments for both and fall back to detecting which category `gdb` on `PATH` falls into if nothing is specified. That would be backwards compatible.

Alternatively we could do auto-detection only and install a good mix of old and new GDB versions on the build bots. 

What do you think, @alexcrichton?

michaelwoerister commented:
> Specifically, in rust mode, for the command print 'basic_types_globals::I', the '...' is interpreted as an unterminated char literal, and I don't know how to fix this.

For me, just specifying `I` (i.e. the unqualified name of the global) seems to work in the new version.

TimNN commented:
> > Specifically, in rust mode, for the command print 'basic_types_globals::I', the '...' is interpreted as an unterminated char literal, and I don't know how to fix this.
> For me, just specifying I (i.e. the unqualified name of the global) seems to work in the new version.

Ok, I can try that.

TimNN commented:
> > Specifically, in rust mode, for the command print 'basic_types_globals::I', the '...' is interpreted as an unterminated char literal, and I don't know how to fix this.

> For me, just specifying I (i.e. the unqualified name of the global) seems to work in the new version.

That seems to have worked, I updated the two tests where I encountered this situation (`basic-types-(mut)-globals`) to work without `set language c`.

TimNN commented:
I've started removing some of the `set syntax c` lines. While doing so, I have tagged several places as `// FIXME` where RGDB behaves suboptimal (`type_names::&Trait2<...>`) or outright bad (`$4 = <error reading variable>` for univariant enums).

cc @tromey, @Manishearth on the last issue.

TimNN commented:
I've been removing some more `set syntax c` lines, and found two more `FIXME`'s, I'm gonna start tracking all these in the top post.

Manishearth commented on src/test/debuginfo/borrowed-enum.rs:
```diff
@@ -18,13 +18,17 @@
-// gdb-check:$3 = {{__0 = 4820353753753434}}
+// gdbg-check:$3 = {{__0 = 4820353753753434}}
+// FIXME
+// gdbr-check:$3 = <error reading variable>
```
This seems fixed. Trying on trunk in a moment, but this works from a gdb I have from august.

```
(gdb) p univariant_ref
$1 = (borrowed_enum::Univariant *) 0x7fff5fbff820
(gdb) ptype univariant_ref
type = union borrowed_enum::Univariant {
    borrowed_enum::TheOnlyCase;
} *
```

TimNN replied:
Ok, I'm currently working from the GDB that ships with Ubuntu Yakkety Yak (`GNU gdb (Ubuntu 7.11.90.20161005-0ubuntu1) 7.11.90.20161005-git`). I'll see about updating to a newer version tomorrow.

Manishearth replied:
Try and build trunk, and let me know which bugs persist.

TimNN replied:
@Manishearth: I tried again with a trunk build (`GNU gdb (GDB) 7.12.50.20161027-git`) and got four different failures (in `pretty` tests, I see `warning: Unsupported auto-load script at offset 0 in section .debug_gdb_scripts` but haven't investigated further yet).

Also, I noticed, that in your snippet you did not dereference the `univariant_ref`, is it possible that the `<error reading variable>` occurs only when dereferencing?

Manishearth commented:
Yeah, the univariant thing is a bug. Fixed in https://sourceware.org/ml/gdb-patches/2016-10/msg00778.html. Try applying that patch and let me know how many errors are left, it should fix both the univariant issues.

TimNN commented:
@Manishearth: I'll try the patch ~~tomorrow~~ later today.

The trait issue doesn't seem that important to me - sure, it doesn't look great but it's still easy to understand. So if there's an easy fix / you've got the time, go for it, otherwise I wouldn't bother and wait for "trait object support".

Manishearth commented:
The fix would be to print it as `&Trait` instead of `struct type_name::&Trait`, but that's not a major difference and really it kind of _is_ `struct type_name::&Trait` (internally).  I think `type_name::&Trait` is a string generated by rustc, too. I'll wait for trait object support, though if you want this fixed earlier, see if you can make rustc generate the right string.

TimNN commented:
@Manishearth: the patch worked flawlessly, thanks!

There are currently two issues left (although I've not yet gone through all the tests which were "fixed" by simply adding `set language c` at the top):

* [ ] Unions are printed as `<error reading variable>`
* [ ] Before the initial `run` command, gdb does not realise it should be in rust mode (thus, if one want's to print statics before the run one has to explicitly `set language rust`) (maybe this is intended / expected behaviour, if so, someone please tell me ;) (see https://github.com/rust-lang/rust/pull/37410/files#diff-314f334bf26a5c3f111b2d20aba1f3ed)

TimNN commented:
I went through all the tests to which I previously added `set language c` and was able to remove all but two instances of `set language c` (see the FIXME section in the top post).

@Manishearth: at this point I'm most interested in whether it is possible to treat an enum value as being a specific variant in an expressions (see the [`recursive-struct` test](https://github.com/rust-lang/rust/pull/37410/files#diff-aacc384954aacf191891adc7ebaa24a9)).

Manishearth commented:
You can access enum values already. Pretend the type is the struct (in case of a struct variant) or tuple struct corresponding to the variant and use named or numbered field access. E.g. use `some.0` to get the innards of a `Some`. 

That test uses arrow operators btw, those will not work for rust and you have to explicitly deref.

alexcrichton commented:
> Alternatively we could do auto-detection only and install a good mix of old and new GDB versions on the build bots. 

My *absolutely ideal* world would look like:

* The configure scripts auto-detects what gdb is being used (somehow)
* The configure script also has a flag along the lines of `--assert-pre-rust-gdb`
* The bots pass these flags appropriately on two bots

It may unfortunately take awhile to get the bots online and working here, but I don't think that should block this PR. We should land this ahead of that and then flesh out the bugs once we get a real builder up and running.

TimNN commented:
@Manishearth: I think I understand what you said however now the following does not seem to make much sense:

```
(gdb) print stack_unique.next.val
Attempting to access named field val of tuple variant recursive_struct::Opt<Box<recursive_struct::UniqueNode<u16>>>::Val, which has only anonymous fields
(gdb) print stack_unique.next.0
Variant recursive_struct::Opt<Box<recursive_struct::UniqueNode<u16>>>::Val is not a tuple variant
(gdb)
```

The two error messages seem contradictory to me.

This is from the `recursive-struct` test (without `set language c`) but I've extracted the relevant definitions here:

```rust
enum Opt<T> {
    Empty,
    Val { val: T }
}

struct UniqueNode<T> {
    next: Opt<Box<UniqueNode<T>>>,
    value: T
}


fn main() {
    let stack_unique: UniqueNode<u16> = UniqueNode {
        next: Val {
            val: box UniqueNode {
                next: Empty,
                value: 1,
            }
        },
        value: 0,
    };
}
```

**Edit:** The problem goes away, when adding a third variant to `Opt`, so I guess the problem is with option like / structs (which have non-zero / pointer optimizations).

Manishearth commented:
Hm, that sounds like a bug.

Manishearth commented:
> "Nil Enums" (`enum Foo {}`) are printed as `<error reading variable>` (probably expected? I'm not sure what to make of that test)

These aren't supposed to exist as values. GDB throws an error in this case (not sure why the error text isn't shown, that might be a bug)

>  Unions are printed as `<error reading variable>`

There's no support for these yet (they didn't exist when we did the rust support). I'll try to add it later.


Regarding `set language rust`, I think there was an issue at one point with gdb not autodetecting this but it should work now. If this still happens I think it's a bug.

Manishearth commented:
```patch
diff --git a/gdb/rust-lang.c b/gdb/rust-lang.c
index 9d13353..9569584 100644
--- a/gdb/rust-lang.c
+++ b/gdb/rust-lang.c
@@ -1736,7 +1736,9 @@ tuple structs, and tuple-like enum variants"));
 	    variant_type = TYPE_FIELD_TYPE (type, disr.field_no);
 
 	    if (variant_type == NULL
-		|| rust_tuple_variant_type_p (variant_type))
+	        || (disr.is_encoded
+	            ? rust_tuple_struct_type_p (variant_type)
+	            : rust_tuple_variant_type_p (variant_type)))
 	      error(_("Attempting to access named field %s of tuple variant %s, \
 which has only anonymous fields"),
 		    field_name, disr.name);
```

This should fix the struct variant issue. Not yet probed properly into this, it could be the wrong fix.

TimNN commented:
> This should fix the struct variant issue.

This did indeed fix my struct variant issues.

> That test uses arrow operators btw, those will not work for rust and **you have to explicitly deref**.

(Emphasis mine) Just, FYI, I found that explicitly dereferencing was not necessary, just using the dot operator was sufficient.

TimNN commented:
> >"Nil Enums" (enum Foo {}) are printed as <error reading variable> (probably expected? I'm not sure what to make of that test)
> These aren't supposed to exist as values. GDB throws an error in this case (not sure why the error text isn't shown, that might be a bug)

Yeah, I thought so, which is why I was confused that this test even exists.

> Regarding `set language rust`, I think there was an issue at one point with gdb not autodetecting this but it should work now. If this still happens I think it's a bug. IIRC last time it was a rustc bug (not emitting the right dwarf language tag). It works for me on last week's nightly + gdb master.

`set language rust` is still required in to places on GDB master. Note that as soon as the gdb `run` command is executed, gdb correctly detects the language as rust. The problem only occurs when inspecting statics before executing `run`.

TimNN commented:
Since the test have all been updated now, I'm gonna come back on the gdb version handling:

**auto-detection**

No matter what approach we choose it seems as if auto-detecting is desirable / required. Since @michaelwoerister expressed concerns about the stability of the error message (which I verified: gdb 6.6 for example, has a different error message) we should choose a version based approach.

I'm not familiar with the versioning scheme used by gdb (what's the relation between 7.11.1 and 7.11.10?) but based on some guessing I would suggest treating gdb >= 7.11.10 as having rust support.

**build system**

@alexcrichton and @michaelwoerister have expressed somewhat contrary opinions on this point. Irrespective of which approach we implement, I propose
* to have the gdb-kind detection in compiletest (otherwise we'll have to duplicate that the logic between rustbuild & configure)
* to add an option to compiletest, rustbuild & configure to specify the gdb executable (defaulting to `gdb`)

Which leaves us with two other requests (which may *both* be implemented):

* @michaelwoerister expressed interest in having one bot test two gdb versions. To accomplish this I propose adding a `gdb-extra` option to rustbuild & configure, which defaults to "unset" and, if set, causes compiletest to be invoked twice, once for `gdb` and once for `gdb-extra`.
* @alexcrichton expressed interest in asserting the kind of gdb. To accomplish this I propose adding a `assert-gdb-kind=pre-rust|post-rust` option to rustbuild, configure & compiletest, defaulting to "unset". If this option is set, compiletest will validate the assertion. This has the drawback of delaying the assertion until compiletest is run. (An analogous option could be added for `gdb-extra`.

Manishearth commented:
> I have found not way of interpreting a value as a specific enum variant, so had to resort to set language c

Is this still the case?

Manishearth commented:
Here's untagged unions btw

<details>
```patch
diff --git a/gdb/rust-lang.c b/gdb/rust-lang.c
index 9569584..5376efc 100644
--- a/gdb/rust-lang.c
+++ b/gdb/rust-lang.c
@@ -93,6 +93,28 @@ struct disr_info
 
 #define RUST_ENCODED_ENUM_HIDDEN 1
 
+/* Whether or not a TYPE_CODE_UNION value is an untagged union
+   as opposed to being a regular Rust enum.  */
+static bool
+rust_union_is_untagged(struct type *type) {
+  /* Unions must have at least one field.  */
+  if (TYPE_NFIELDS (type) == 0)
+    return false;
+  /* If the first field is named, but the name has the rust enum prefix,
+      it is an enum.  */
+  if (strncmp (TYPE_FIELD_NAME (type, 0), RUST_ENUM_PREFIX,
+         strlen (RUST_ENUM_PREFIX)) == 0)
+    return false;
+  /* Unions only have named fields.  */
+  for (int i = 0; i < TYPE_NFIELDS (type); ++i)
+    {
+      if (strlen (TYPE_FIELD_NAME (type, i)) == 0)
+        return false;
+    }
+    return true;
+
+}
+
 /* Utility function to get discriminant info for a given value.  */
 
 static struct disr_info
@@ -566,6 +588,14 @@ rust_val_print (struct type *type, const gdb_byte *valaddr, int embedded_offset,
 	struct value_print_options opts;
 	struct cleanup *cleanup;
 
+  /* Untagged unions are printed as if they are structs.
+     Since the field bit positions overlap in the debuginfo,
+     the code for printing a union is same as that for a struct,
+     the only difference is that the input type will have overlapping
+     fields.  */
+  if (rust_union_is_untagged (type))
+      goto struct_val;
+
 	opts = *options;
 	opts.deref_ref = 0;
 
@@ -638,6 +668,7 @@ rust_val_print (struct type *type, const gdb_byte *valaddr, int embedded_offset,
       break;
 
     case TYPE_CODE_STRUCT:
+    struct_val:
       {
 	int i;
 	int first_field;
@@ -809,6 +840,7 @@ rust_print_type (struct type *type, const char *varstring,
       break;
 
     case TYPE_CODE_STRUCT:
+    struct_printer:
       {
 	int is_tuple_struct;
 
@@ -823,7 +855,12 @@ rust_print_type (struct type *type, const char *varstring,
 	if (TYPE_N_BASECLASSES (type) > 0)
 	  goto c_printer;
 
-	fputs_filtered ("struct ", stream);
+  /* This code path is also used by unions.  */
+  if (TYPE_CODE (type) == TYPE_CODE_STRUCT)
+	  fputs_filtered ("struct ", stream);
+  else
+    fputs_filtered ("union ", stream);
+
 	if (TYPE_TAG_NAME (type) != NULL)
 	  fputs_filtered (TYPE_TAG_NAME (type), stream);
 
@@ -899,6 +936,13 @@ rust_print_type (struct type *type, const char *varstring,
 	/* Skip the discriminant field.  */
 	int skip_to = 1;
 
+  /* Unions and structs have the same syntax in Rust,
+     the only difference is that structs are declared with `struct`
+     and union with `union`. This difference is handled in the struct
+     printer.  */
+  if (rust_union_is_untagged (type))
+    goto struct_printer;
+
 	fputs_filtered ("enum ", stream);
 	if (TYPE_TAG_NAME (type) != NULL)
 	  {
@@ -1637,7 +1681,10 @@ rust_evaluate_subexp (struct type *expect_type, struct expression *exp,
         lhs = evaluate_subexp (NULL_TYPE, exp, pos, noside);
 
         type = value_type (lhs);
-        if (TYPE_CODE (type) == TYPE_CODE_UNION)
+        /* Untagged unions can't have anonymous field access since
+           they can only have named fields.  */
+        if (TYPE_CODE (type) == TYPE_CODE_UNION
+            && !rust_union_is_untagged (type))
 	  {
 	    struct cleanup *cleanup;
 
@@ -1712,8 +1759,8 @@ tuple structs, and tuple-like enum variants"));
         lhs = evaluate_subexp (NULL_TYPE, exp, pos, noside);
 
         type = value_type (lhs);
-
-        if (TYPE_CODE (type) == TYPE_CODE_UNION)
+        if (TYPE_CODE (type) == TYPE_CODE_UNION
+            && !rust_union_is_untagged (type))
 	  {
 	    int i, start;
 	    struct disr_info disr;
@@ -1762,6 +1809,7 @@ which has only anonymous fields"),
 	  }
 	else
 	  {
+      /* Field access in structs and untagged unions works like C.  */
 	    *pos = pc;
 	    result = evaluate_subexp_standard (expect_type, exp, pos, noside);
 	  }
```
</details>

All three patches have been submitted upstream (and can also be found at https://github.com/manishearth/gdb/tree/enum-union-fixes), just need review.

Manishearth commented:
I wonder if we should just test on gdb trunk on CI? Check the hash in-tree, and update it whenever GDB rust support is improved. GDB itself has some tests for its rust output, but they're not rigorously run and can be broken by rust updates.

Even better, have both gdb trunk and a gdb release, and when they differ have two lines in the debuginfo tests.

alexcrichton commented:
> but based on some guessing I would suggest treating gdb >= 7.11.10 as having rust support

:+1:

> expressed interest in having one bot test two gdb versions

I would personally be ok punting on this for now. We should get *a* bot testing the Rust GDB first, and then once we get that up and running I figure we can learn how to test both GDBs at once. @michaelwoerister what do you think though? I agree ideally we'd have one bot test *all* the GDB versions, however.

> To accomplish this I propose adding a assert-gdb-kind=pre-rust|post-rust option to rustbuild, configure & compiletest, defaulting to "unset". If this option is set, compiletest will validate the assertion

Sounds good to me! I'm ok with the minor drawback here.

Manishearth commented:
> The problem only occurs when inspecting statics before executing run.

I played with this a bit more. At first I thought it was a gdb bug but it worked fine with Go binaries.

Turns out that gdb is kinda-sorta correct in thinking this is a C program. It is! jemalloc has a bunch of things statically linked in, and those are all in C, with the appropriate debug header. gdb seems to consider the last seen `AT_language` tag as the toplevel language. Perhaps it should consider the first instead; not sure if that's the right fix (@tromey?)

TimNN commented:
> > I have found not way of interpreting a value as a specific enum variant, so had to resort to `set language c`

> Is this still the case?

No, thats working now. The only place left where I'm currently using `set language c` is in the [`packed-struct` test](https://github.com/rust-lang/rust/pull/37410/files#diff-d26ee700e8c5945ed55e748863bcfe67) because it uses `sizeof` to verify the size of some values. Is there an equivalent thing we can do in rust mode?

TimNN commented:
> Here's untagged unions btw

Awesome! Seems to be working great (although the test is not really exhaustive).

TimNN commented:
> > I have found not way of interpreting a value as a specific enum variant, so had to resort to `set language c`

> Is this still the case?

No, thats working now. The only place left where I'm currently using `set language c` is 

---

I'll implement the build system changes over the next few days (and am also going to squash all those test updating commits, as long as nobody objects).

Manishearth commented:
Added a sizeof operator in the form of `sizeof(foo)` (patch pushed to the same branch):

<details>
diff --git a/gdb/rust-exp.y b/gdb/rust-exp.y
index 6dc4704..dffccd0 100644
--- a/gdb/rust-exp.y
+++ b/gdb/rust-exp.y
@@ -276,6 +276,7 @@ struct rust_op
 %token <voidval> KW_EXTERN
 %token <voidval> KW_CONST
 %token <voidval> KW_FN
+%token <voidval> KW_SIZEOF
 
 /* Operator tokens.  */
 %token <voidval> DOTDOT
@@ -371,7 +372,7 @@ expr:
 |	array_expr
 |	idx_expr
 |	range_expr
-|	unop_expr
+|	unop_expr /* Must precede call_expr because of ambiguity with sizeof.  */
 |	binop_expr
 |	paren_expr
 |	call_expr
@@ -577,7 +578,8 @@ unop_expr:
 
 |	'&' KW_MUT expr	%prec UNARY
 		{ $$ = ast_unary (UNOP_ADDR, $3); }
-
+|   KW_SIZEOF '(' expr ')' %prec UNARY
+        { $$ = ast_unary (UNOP_SIZEOF, $3); }
 ;
 
 binop_expr:
@@ -872,6 +874,7 @@ static const struct token_info identifier_tokens[] =
   { "true", KW_TRUE, OP_NULL },
   { "extern", KW_EXTERN, OP_NULL },
   { "fn", KW_FN, OP_NULL },
+  { "sizeof", KW_SIZEOF, OP_NULL },
 };
 
 /* Operator tokens, sorted longest first.  */
@@ -2194,6 +2197,7 @@ convert_ast_to_expression (struct parser_state *state,
     case UNOP_COMPLEMENT:
     case UNOP_IND:
     case UNOP_ADDR:
+    case UNOP_SIZEOF:
       convert_ast_to_expression (state, operation->left.op, top);
       write_exp_elt_opcode (state, operation->opcode);
       break;
diff --git a/gdb/testsuite/gdb.rust/simple.exp b/gdb/testsuite/gdb.rust/simple.exp
index 075bff7..9ab4b70 100644
--- a/gdb/testsuite/gdb.rust/simple.exp
+++ b/gdb/testsuite/gdb.rust/simple.exp
@@ -33,6 +33,7 @@ if {![runto ${srcfile}:$line]} {
 
 gdb_test "print a" " = \\(\\)"
 gdb_test "ptype a" " = \\(\\)"
+gdb_test "print sizeof(a)" " = 0"
 
 gdb_test "print b" " = \\\[\\\]"
 gdb_test "ptype b" " = \\\[i32; 0\\\]"
@@ -41,6 +42,7 @@ gdb_test "print *(&b as *const \[i32; 0_0\])" " = \\\[\\\]"
 
 gdb_test "print c" " = 99"
 gdb_test "ptype c" " = i32"
+gdb_test "print sizeof(c)" " = 4"
 
 gdb_test "print c = 87" " = \\(\\)"
 gdb_test "print c" " = 87"
@@ -124,7 +126,7 @@ gdb_test "print nosuchsymbol" \
 gdb_test "print e" " = simple::MoreComplicated::Two\\(73\\)"
 gdb_test "print e2" \
     " = simple::MoreComplicated::Four\\{this: true, is: 8, a: 109 'm', struct_: 100, variant: 10\\}"
-
+gdb_test "print sizeof(e)" " = 24"
 gdb_test_sequence "ptype e" "" {
     " = enum simple::MoreComplicated \\{"
     "  One,"
</details>

TimNN commented:
> Added a sizeof operator in the form of sizeof(foo) (patch pushed to the same branch):

@Manishearth: Awesome, works like a charm!

michaelwoerister commented:
> I would personally be ok punting on this for now. We should get a bot testing the Rust GDB first, and then once we get that up and running I figure we can learn how to test both GDBs at once. @michaelwoerister what do you think though? I agree ideally we'd have one bot test all the GDB versions, however.

Yes, testing both GDB versions on all bots is something I consider a long-term goal, not something that needs to happen with PR.

alexcrichton commented:
@TimNN is this ready to go now modulo getting a bot up and running? If so it looks good to r+ to me, and we can work on a bot afterwards.

TimNN commented:
@alexcrichton: The `assert-gdb-kind` flag is still missing, although I could add that in a separate PR.

alexcrichton commented:
Ah yeah that's ok, we can add that once we start gating as well. Other than that though we're good to go?

TimNN commented:
Yeah, I think so.

alexcrichton commented:
@bors: r+ e72e21433defc193b9896a79dcbb6535ee05ee85

bors commented:
:hourglass: Testing commit e72e214 with merge 293b25f...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/2185)

TimNN commented:
Alright, so I think I fixed the problem:

* `./configure` determined an absolute path for `gdb`
* bootstrap attempted to resolve this path against `$PATH` (which of course failed)
* to fix this, I added a heuristic to `have_cmd` to determine whether we are given a path or only a command name (see the last commit).

alexcrichton commented:
Hm I think this is an MSYS2-ism leaking a bit too far into bootstrap. I think what we may want to do instead is to add support like [the function in another pending PR](https://github.com/rust-lang/rust/pull/36986/files#diff-5330a247a9bddb2afb405bbf4d279c1bR33) which converts MSYS2 paths to Windows-style paths. That is, the actual component at fault here is that when we read variables from config.mk we don't properly translate them.

I think that if a tool is an absolute path that should work because `Path::join` is just a replacement operation in that case.

tromey commented:
> Turns out that gdb is kinda-sorta correct in thinking this is a C program. It is! jemalloc has a bunch of things statically linked in, and those are all in C, with the appropriate debug header. gdb seems to consider the last seen AT_language tag as the toplevel language. Perhaps it should consider the first instead; not sure if that's the right fix (@tromey?)

gdb generally uses the language of the current frame.  If there are no frames (i.e., at startup), then it takes the language of `main` (actually a "static location", see below).  The debuginfo reader has an opportunity to tell gdb some other name (and language) for `main`.  Ideally this would be done per https://github.com/rust-lang/rust/issues/32620 (gdb doesn't actually handle `DW_AT_main_subprogram` presently, but I'd be happy to implement it if rustc were to emit it).

The "static location" can also be changed by commands like `list`.  But on the whole I think `set language rust` is clearer.

TimNN commented:
> Hm I think this is an MSYS2-ism leaking a bit too far into bootstrap. I think what we may want to do instead is to add support like the function in another pending PR which converts MSYS2 paths to Windows-style paths. That is, the actual component at fault here is that when we read variables from config.mk we don't properly translate them.

Sounds good! If there are no objections I'm gonna apply this transformation to all paths read from `config.mk`.

> I think that if a tool is an absolute path that should work because Path::join is just a replacement operation in that case.

Ah good point, forgot about that (and was wondering, why absolute paths were working on linux...)

> But on the whole I think `set language rust` is clearer.

Thanks for the feedback @tromey.

nikomatsakis commented:
r? @michaelwoerister -- he's more qualified to review this than me

alexcrichton commented:
> If there are no objections I'm gonna apply this transformation to all paths read from config.mk.

👍 j

TimNN commented:
I updated the PR - r? @alexcrichton

alexcrichton commented on src/bootstrap/config.rs:
```diff
@@ -417,51 +417,74 @@ impl Config {
+        return path.into();
+    }
+
+    let win_path = build_helper::output(&mut Command::new("cygpath").arg("-w").arg(path));
```
This extra `&mut` I don't think is necessary (returned by `arg`)

alexcrichton commented on src/bootstrap/config.rs:
```diff
@@ -449,8 +450,11 @@ impl Config {
-                    self.rustc = Some(PathBuf::from(value).join("bin/rustc"));
-                    self.cargo = Some(PathBuf::from(value).join("bin/cargo"));
+                    let rustc = if cfg!(windows) { "rustc.exe" } else { "rustc" };
+                    let cargo = if cfg!(windows) { "cargo.exe" } else { "cargo" };
```
Perhaps this could be a helper in the `util` module to reduce duplication?

TimNN commented:
@alexcrichton: Updated.

alexcrichton commented:
@bors: r+

Thanks @TimNN!

bors commented:
:pushpin: Commit f7107f3 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton f7107f32c2fabef7a5b33c514d9f0310c96492d0 -->

bors commented:
:hourglass: Testing commit f7107f3 with merge 8e7bfe5...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2869)

michaelwoerister commented:
@bors retry

bors commented:
:hourglass: Testing commit f7107f3 with merge c88514d...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2880)

alexcrichton commented:
@bors: retry

Manishearth commented:
GDB fixes landed in trunk: [1](https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=commit;h=51a789c3bf3d9b04d3d81493fda7f2514ae43add), [2](https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=commit;h=b96645f1a1e6084ed6a20ddf615bdc4d2487a9c9), [3](https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=commit;h=cdf5a07c75b47481da1e99ee91ee860c2a6bb5f6).

bors commented:
:hourglass: Testing commit f7107f3 with merge 9d34708...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/2237)

TimNN commented:
I fixed the android failure (https://github.com/rust-lang/rust/pull/37597#issuecomment-258597679).

alexcrichton commented:
@TimNN awesome thanks! I'll update the rollup

alexcrichton commented:
@bors: r+ force clean

bors commented:
:pushpin: Commit d2cb515 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton d2cb515ab045eefa1fb7ec36078fa0440e4affec -->

bors commented:
:hourglass: Testing commit d2cb515 with merge 366636e...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2951)

TimNN commented:
I fixed the windows failure (although that seems to have been fixed in the rollup already).

alexcrichton commented:
Hm I must have botched the merge but this was landed in https://github.com/rust-lang/rust/pull/37597, so closing.

michaelwoerister commented:
🎉
